modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C

```diff
@@ -64,72 +52,63 @@ SamplerPostprocessorTransfer::initialSetup()
   if (!_results)
     mooseError("The 'results' object must be a 'StochasticResults' object.");
 
-  _results->init(*_sampler);
+  _results->init(*_sampler_ptr);
 }
 
 void
 SamplerPostprocessorTransfer::initializeFromMultiapp()
 {
-  _local_values.clear();
+  VectorPostprocessorValue & vpp =
+      _fe_problem.getVectorPostprocessorValue(_master_vpp_name, _sampler_ptr->name(), false);
+
+  if (_results->containsCompleteHistory())
+    vpp.reserve(vpp.size() + _sampler_ptr->getNumberOfLocalRows());
+  else
+  {
+    vpp.clear();
+    vpp.reserve(_sampler_ptr->getNumberOfLocalRows());
+  }
 }
 
 void
 SamplerPostprocessorTransfer::executeFromMultiapp()
 {
+  VectorPostprocessorValue & vpp =
+      _fe_problem.getVectorPostprocessorValue(_master_vpp_name, _sampler_ptr->name(), false);
+
   const dof_id_type n = _multi_app->numGlobalApps();
   for (MooseIndex(n) i = 0; i < n; i++)
   {
     if (_multi_app->hasLocalApp(i))
     {
       FEProblemBase & app_problem = _multi_app->appProblemBase(i);
-      _local_values.push_back(app_problem.getPostprocessorValue(_sub_pp_name));
+      vpp.push_back(app_problem.getPostprocessorValue(_sub_pp_name));
     }
   }
 }
 
 void
 SamplerPostprocessorTransfer::finalizeFromMultiapp()
 {
-  // Gather the PP values from all ranks
-  _communicator.gather(0, _local_values);
```

Comment:
should we still need the gather here? otherwise we only send the postprocessor on each processor to the vector postprocessor?

Reply:
That is the purpose; the StochasticResults performs the gather if needed.